ebpf_extensions/neteventebpfext/netevent_ebpf_ext_event.c

```diff
@@ -0,0 +1,530 @@
+static ebpf_extension_hook_provider_t* _ebpf_netevent_event_hook_provider_context = NULL;
+EX_PUSH_LOCK _ebpf_netevent_event_hook_provider_lock;
+bool _ebpf_netevent_event_hook_provider_registered = FALSE;
+uint64_t _ebpf_netevent_event_hook_provider_registration_count = 0;
+
+//
+// Event Program Information NPI Provider.
+//
+static ebpf_program_data_t _ebpf_netevent_event_program_data = {
+    .header = {.version = EBPF_PROGRAM_DATA_CURRENT_VERSION, .size = EBPF_PROGRAM_DATA_CURRENT_VERSION_SIZE},
+    .program_info = &_ebpf_netevent_event_program_info,
+    .program_type_specific_helper_function_addresses = NULL, // No helper functions exposed to client eBPF programs.
+    .context_create = _ebpf_netevent_program_context_create,
+    .context_destroy = _ebpf_netevent_program_context_destroy,
+    .required_irql = PASSIVE_LEVEL,
+};
+static ebpf_extension_data_t _ebpf_netevent_event_program_info_provider_data = {
+    EBPF_EXTENSION_NPI_PROVIDER_VERSION, sizeof(_ebpf_netevent_event_program_data), &_ebpf_netevent_event_program_data};
+
+NPI_MODULEID DECLSPEC_SELECTANY _ebpf_netevent_event_program_info_provider_moduleid = {
+    sizeof(NPI_MODULEID), MIT_GUID, {0}};
+
+static ebpf_extension_program_info_provider_t* _ebpf_netevent_event_program_info_provider_context = NULL;
+
+//
+// Event Hook NPI Client Attach and Detach Callbacks (to NetEvent NPI provider).
+// Callbacks invoked when a Program Information NPI client attaches/detaches.
+// (must register a separate Hook NPI provider module for each supported attach type)
+//
+static ebpf_result_t
+_netevent_ebpf_extension_netevent_event_on_client_attach(
+    _In_ const ebpf_extension_hook_client_t* attaching_client,
+    _In_ const ebpf_extension_hook_provider_t* provider_context)
+{
+    ebpf_result_t result = EBPF_SUCCESS;
+    bool push_lock_acquired = false;
+
+    EBPF_EXT_LOG_ENTRY();
+
+    UNREFERENCED_PARAMETER(attaching_client);
+    UNREFERENCED_PARAMETER(provider_context);
+
+    ExAcquirePushLockExclusive(&_ebpf_netevent_event_hook_provider_lock);
+
+    push_lock_acquired = true;
+
+    if (!_ebpf_netevent_event_hook_provider_registered) {
+        // Register and attach the neteventebpfext extension to NetEvent as an NMR Client.
+        // This will invoke the _netevent_ebpf_extension_attach_provider() callback.
+        NTSTATUS status = NmrRegisterClient(
+            &_netevent_client_characteristics,
+            &_netevent_client_registration_context,
+            &_netevent_client_binding_context.nmr_binding_handle);
+        if (!NT_SUCCESS(status)) {
+            EBPF_EXT_LOG_MESSAGE_NTSTATUS(
+                EBPF_EXT_TRACELOG_LEVEL_ERROR, EBPF_EXT_TRACELOG_KEYWORD_NETEVENT, "Attach to netevent failed", status);
+            result = EBPF_OPERATION_NOT_SUPPORTED;
+            goto Exit;
+        }
+        _ebpf_netevent_event_hook_provider_registered = TRUE;
+    }
+
+    _ebpf_netevent_event_hook_provider_registration_count++;
+
+Exit:
+    if (push_lock_acquired) {
```

Comment:
This is a safety coding pattern (btw used in the ntosext extension as well), to ensure that any future modification will enforce to correctly handle the lock release. Although in this specific case it may be simplified, it is best the have the pattern in place, given its near to zero cost.